File: 1/2021.07.27/557.Reverse Words in a String III/python/solution.py
class Solution:
    def reverseWords(self, s: str) -> str:
        def reverse(start, end):
            while start < end:
                cs[start], cs[end] = cs[end], cs[start]
                start += 1
                end -= 1
        
        cs = list(s)
        i = 0
        while i < len(s):
            j = i + 1
            while j < len(s):
                if cs[j] == ' ':
                    break
                else:
                    j += 1
            reverse(i, j - 1)
            i = j + 1
        return ''.join(cs)

File: 1/2021.07.27/557.Reverse Words in a String III/java/solution.java
class Solution {
    public String reverseWords(String s) {
        char[] cs = s.toCharArray();
        int i = 0;
        while (i < s.length()) {
            int j = i + 1;
            while (j < s.length()) {
                if (cs[j] == ' ') {
                    break;
                } else {
                    j++;
                }
            }
            reverse(cs, i, j - 1);
            i = j + 1;
        }
        return new String(cs);
    }

    private void reverse(char[] cs, int start, int end) {
        while (start < end) {
            char tmp = cs[start];
            cs[start] = cs[end];
            cs[end] = tmp;
            start++;
            end--;
        }
    }
}
